arch/x86/real_prot_dance.S

```diff
@@ -79,24 +82,75 @@ ENTRY(_long_to_real)
     /* Save stack pointer */
     mov   %rsp, _saved_sp
 
+    movb  $RETURN_TO_LONG_MODE, return_mode
+
     push  $__KERN_CS32
-    push  $.Lprot_mode
+    push  $.Lfrom_long_mode
     lretq
 END_FUNC(_long_to_real)
 
 .code32
-.align 16
-.Lprot_mode:
-    /* Load real mode accessible GDT and IDT */
-    lgdt  rmode_gdt_ptr
-    lidt  rmode_idt_ptr
+ENTRY(_prot_to_real)
+    /* Save current GDT and IDT in low memory area */
+    mov   4(%esp), %eax
+    mov   %eax, _saved_idt_ptr
+    mov   8(%esp), %eax
+    mov   %eax, _saved_gdt_ptr
+
+    /* Save current flags register */
+    pushfl
+
+    /* Disable interrupts. If we get NMI or MC, hope we get back here */
+    cli
```

Comment:
Can you re-enable them after the setup?

Reply:
The current state is saved with the `pushfl` on line 101. Upon return, the `popfl` restores the setting.